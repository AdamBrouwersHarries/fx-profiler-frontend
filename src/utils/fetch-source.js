/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { assertExhaustiveCheck } from './flow';
import {
  getDownloadRecipeForSourceFile,
  parseFileNameFromSymbolication,
} from './special-paths';
import type { SourceLoadingError, AddressProof } from 'firefox-profiler/types';

export type FetchSourceCallbacks = {|
  fetchUrlResponse: (url: string, postData?: MixedObject) => Promise<Response>,
|};

export type FetchSourceResult =
  | { type: 'SUCCESS', source: string }
  | { type: 'ERROR', errors: SourceLoadingError[] };

/**
 * Fetch the source code for a file path from the web.
 *
 * For example, if `file` is "hg:hg.mozilla.org/mozilla-central:widget/cocoa/nsAppShell.mm:997f00815e6bc28806b75448c8829f0259d2cb28",
 * then this will call `callbacks.fetchUrlResponse("https://hg.mozilla.org/mozilla-central/raw-file/997f00815e6bc28806b75448c8829f0259d2cb28/widget/cocoa/nsAppShell.mm")`.
 */
export async function fetchSource(
  file: string,
  symbolServerUrl: string,
  addressProof: AddressProof | null,
  callbacks: FetchSourceCallbacks
): Promise<FetchSourceResult> {
  const errors: SourceLoadingError[] = [];

  // See if we can get sources from a local symbol server first.
  if (_serverMightSupportSource(symbolServerUrl) && addressProof !== null) {
    const { debugName, breakpadId, address } = addressProof;

    // Make a request to /source/v1. The API format for this endpoint is documented
    // at https://github.com/mstange/profiler-get-symbols/blob/master/API.md#sourcev1
    const url = `${symbolServerUrl}/source/v1`;
    const body = JSON.stringify({
      debugName,
      debugId: breakpadId,
      moduleOffset: `0x${address.toString(16)}`,
      file,
    });

    try {
      const response = await callbacks.fetchUrlResponse(url, body);
      const responseJson = await response.json();
      if (!responseJson.error) {
        // Local symbol server gave us the source. Success!
        return { type: 'SUCCESS', source: responseJson.source };
      }
      errors.push({
        type: 'SYMBOL_SERVER_API_ERROR',
        apiErrorMessage: responseJson.error,
      });
    } catch (e) {
      errors.push({
        type: 'NETWORK_ERROR',
        url,
        networkErrorMessage: e.toString(),
      });
    }
  }

  // We did not get the source from a local symbol server. Try to obtain it
  // from the web instead.

  const parsedName = parseFileNameFromSymbolication(file);
  const downloadRecipe = getDownloadRecipeForSourceFile(parsedName);

  switch (downloadRecipe.type) {
    case 'CORS_ENABLED_SINGLE_FILE': {
      const { url } = downloadRecipe;
      try {
        const response = await callbacks.fetchUrlResponse(url);
        const source = await response.text();
        return { type: 'SUCCESS', source };
      } catch (e) {
        errors.push({
          type: 'NETWORK_ERROR',
          url,
          networkErrorMessage: e.toString(),
        });
      }
      break;
    }
    case 'CORS_ENABLED_ARCHIVE': {
      // Not handled yet.
      errors.push({ type: 'NO_KNOWN_CORS_URL' });
      break;
    }
    case 'NO_KNOWN_CORS_URL': {
      errors.push({ type: 'NO_KNOWN_CORS_URL' });
      break;
    }
    default:
      throw assertExhaustiveCheck(downloadRecipe.type);
  }
  return { type: 'ERROR', errors };
}

// At the moment, the official Mozilla symbolication server does not have an
// endpoint for requesting source code. The /source/v1 URL is only supported by
// local symbol servers. Check the symbol server URL to avoid hammering the
// official Mozilla symbolication server with requests it can't handle.
// This check can be removed once it adds support for /source/v1.
function _serverMightSupportSource(symbolServerUrl: string): boolean {
  try {
    const url = new URL(symbolServerUrl);
    return (
      url.hostname === 'localhost' ||
      url.hostname === '127.0.0.1' ||
      url.hostname === '::1'
    );
  } catch (e) {
    return false;
  }
}
