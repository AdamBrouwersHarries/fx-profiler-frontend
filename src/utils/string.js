/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import escapeStringRegexp from 'escape-string-regexp';

// Initializing this RegExp outside of removeURLs because that function is in a
// hot path during sanitization and it's good to avoid the initialization of the
// RegExp which is costly.
const REMOVE_URLS_REGEXP = (function () {
  const protocols = [
    'http',
    'https',
    'ftp',
    'file',
    'moz-extension',
    'moz-page-thumb',
  ];

  return new RegExp(
    `\\b((?:${protocols.join('|')})://)/?[^\\s/$.?#][^\\s)]*`,
    //    ^                              ^          ^
    //    |                              |          Matches any characters except
    //    |                              |          whitespaces and ')' character.
    //    |                              |          Other characters are allowed now
    //    |                              Matches any character except whitespace
    //    |                              and '/', '$', '.', '?' or '#' characters
    //    |                              because this is start of the URL
    //    Matches URL schemes we need to sanitize.
    'gi'
  );
})();

/**
 * Takes a string and returns the string with public URLs removed.
 * It doesn't remove the URLs like `chrome://..` because they are internal URLs
 * and they shouldn't be removed.
 */
export function removeURLs(
  string: string,
  redactedText: string = '<URL>'
): string {
  return string.replace(REMOVE_URLS_REGEXP, '$1' + redactedText);
}

/**
 * Take an absolute file path string and sanitize it except the last file name segment.
 *
 * Note: Do not use this function if the string not only contains a file path but
 * also contains more text. This function is intended to use only for path strings.
 */
export function removeFilePath(
  filePath: string,
  redactedText: string = '<PATH>'
): string {
  let pathSeparator = null;

  // Figure out which separator the path uses and the last separator index.
  let lastSeparatorIndex = filePath.lastIndexOf('/');
  if (lastSeparatorIndex !== -1) {
    // This is a Unix-like path.
    pathSeparator = '/';
  } else {
    lastSeparatorIndex = filePath.lastIndexOf('\\');
    if (lastSeparatorIndex !== -1) {
      // This is a Windows path.
      pathSeparator = '\\';
    }
  }

  if (pathSeparator === null) {
    // There is no path separator, which means it's either not a file path or empty.
    return filePath;
  }

  return redactedText + pathSeparator + filePath.slice(lastSeparatorIndex + 1);
}

/**
 * Divide a search string into several parts by splitting on comma.
 */
export const splitSearchString = (searchString: string): string[] | null => {
  if (!searchString) {
    return null;
  }
  const result = searchString
    .split(',')
    .map((part) => part.trim())
    .filter((part) => part);

  if (result.length) {
    return result;
  }

  return null;
};

/**
 * Concatenate an array of strings into a RegExp that matches on all
 * the strings.
 */
export const stringsToRegExp = (strings: string[] | null): RegExp | null => {
  if (!strings || !strings.length) {
    return null;
  }

  const regexpStr = strings.map(escapeStringRegexp).join('|');
  return new RegExp(regexpStr, 'gi');
};

export type MarkerSearchFieldMap = Map<
  string,
  {| positive: RegExp | null, negative: RegExp | null |},
>;

export type MarkerRegExps = $ReadOnly<{|
  generic: RegExp | null,
  fieldMap: MarkerSearchFieldMap,
|}>;

/**
 * Concatenate an array of strings into multiple RegExps that match on all
 * the marker strings which can include positive and negative field specific search.
 */
export const stringsToMarkerRegExps = (
  strings: string[] | null
): MarkerRegExps | null => {
  if (!strings || !strings.length) {
    return null;
  }

  // We create this map to group all the field specific search strings and then
  // we aggregate them to create a single regexp for each field later.
  const fieldStrings: Map<
    string,
    {| positive: string[], negative: string[] |},
  > = new Map();
  // These are the non-field specific search strings. They have to be positive
  // as we don't support negative generic filtering.
  const genericPositiveStrings = [];
  for (const string of strings) {
    // First capture group is used to determine if it has a "-" in front of the
    // field to understand if it's a negative filter.
    // Second capture group is used to get the field name.
    // Third capture group is to get the filter value.
    const prefixMatch = string.match(
      /^(?<maybeNegative>-?)(?<key>\w+):(?<value>.+)/i
    );
    if (prefixMatch && prefixMatch.groups) {
      // This is a key-value pair that will only be matched for a specific field.
      const { maybeNegative, value } = prefixMatch.groups;
      const key = prefixMatch.groups.key.toLowerCase();
      let fieldStrs = fieldStrings.get(key);
      if (!fieldStrs) {
        fieldStrs = { positive: [], negative: [] };
        fieldStrings.set(key, fieldStrs);
      }

      // First capture group checks if we have "-" in front of the string to see
      // if it's a negative filtering.
      if (maybeNegative.length === 0) {
        fieldStrs.positive.push(value);
      } else {
        fieldStrs.negative.push(value);
      }
    } else {
      genericPositiveStrings.push(string);
    }
  }

  // Now we constructed the grouped arrays. Let's convert them into a map of RegExps.
  const fieldMap: MarkerSearchFieldMap = new Map();
  for (const [field, strings] of fieldStrings) {
    fieldMap.set(field, {
      positive: stringsToRegExp(strings.positive),
      negative: stringsToRegExp(strings.negative),
    });
  }

  return {
    generic: stringsToRegExp(genericPositiveStrings),
    fieldMap,
  };
};
