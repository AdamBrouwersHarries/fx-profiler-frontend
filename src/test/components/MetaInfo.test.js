/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { fireFullClick } from '../fixtures/utils';
import { ensureExists } from '../../utils/flow';
import { Provider } from 'react-redux';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import {
  getProfileFromTextSamples,
  addExtraInfoToProfile,
} from '../fixtures/profiles/processed-profile';
import { storeWithProfile } from '../fixtures/stores';
import { MetaInfoPanel } from '../../components/app/MenuButtons/MetaInfo';

describe('app/MetaInfo', function () {
  function setup(store) {
    const renderResult = render(
      <Provider store={store}>
        <MetaInfoPanel />
      </Provider>
    );

    return {
      ...store,
      ...renderResult,
    };
  }

  function setupForMoreInfoButton(addExtraInfos: boolean = false) {
    const initialProfile = getProfileFromTextSamples('A').profile;
    const profile = addExtraInfos
      ? addExtraInfoToProfile(initialProfile)
      : initialProfile;
    return setup(storeWithProfile(profile));
  }

  describe('More Info section', function () {
    it('opens more info section if clicked', () => {
      const { container } = setupForMoreInfoButton(true);

      expect(container).toMatchSnapshot();

      const summary = ensureExists(container.querySelector('.moreInfoSummary'));

      fireFullClick(summary);
      expect(container).toMatchSnapshot();
    });

    it("doesn't show the more info button if there isn't any extra info", () => {
      const { container } = setupForMoreInfoButton(false);
      expect(container).toMatchSnapshot();

      const summary = container.querySelector('.moreInfoSummary');
      expect(summary).toBeNull();
    });
  });
});
