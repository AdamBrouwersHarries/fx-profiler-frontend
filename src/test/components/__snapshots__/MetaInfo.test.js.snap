// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/MetaInfo More Info section doesn't show the more info button if there isn't any extra info 1`] = `
<div>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Interval:
      </span>
      1ms
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Symbols:
      </span>
      Profile is symbolicated
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      />
      <button
        class="photon-button photon-button-micro"
        type="button"
      >
        Re-symbolicate profile
      </button>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Application
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Name and version:
      </span>
      Firefox
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Platform
  </h2>
  <div
    class="metaInfoSection"
  />
</div>
`;

exports[`app/MetaInfo More Info section opens more info section if clicked 1`] = `
<div>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Interval:
      </span>
      1ms
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Symbols:
      </span>
      Profile is symbolicated
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      />
      <button
        class="photon-button photon-button-micro"
        type="button"
      >
        Re-symbolicate profile
      </button>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Application
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Name and version:
      </span>
      Firefox
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Platform
  </h2>
  <div
    class="metaInfoSection"
  />
  <details
    class="metaInfoRow"
  >
    <summary
      class="moreInfoSummary"
    >
      <button
        class="moreInfoButton photon-button photon-button-default photon-button-micro"
        type="button"
      >
        Show More
      </button>
    </summary>
    <div
      class="moreInfoPart"
    >
      <div>
        <h2
          class="metaInfoSubTitle"
        >
          CPU
        </h2>
        <div
          class="metaInfoSection"
        >
          <div
            class="moreInfoRow"
          >
            <span
              class="metaInfoWideLabel"
            >
              CPU 1
            </span>
            <div
              class="moreInfoValue"
            >
              Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
            </div>
          </div>
        </div>
      </div>
      <div>
        <h2
          class="metaInfoSubTitle"
        >
          Memory
        </h2>
        <div
          class="metaInfoSection"
        />
      </div>
      <div>
        <h2
          class="metaInfoSubTitle"
        >
          Hard Drives
        </h2>
        <div
          class="metaInfoSection"
        >
          <div
            class="moreInfoRow"
          >
            <span
              class="metaInfoWideLabel"
            >
              SSD
            </span>
            <div
              class="moreInfoValue"
            >
              Samsung SSD 850 EVO 500GB
            </div>
          </div>
          <div
            class="moreInfoRow"
          >
            <span
              class="metaInfoWideLabel"
            >
              HDD
            </span>
            <div
              class="moreInfoValue"
            >
              Seagate ST1000LM035-1RK172
            </div>
          </div>
        </div>
      </div>
    </div>
  </details>
</div>
`;

exports[`app/MetaInfo More Info section opens more info section if clicked 2`] = `
<div>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Interval:
      </span>
      1ms
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Symbols:
      </span>
      Profile is symbolicated
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      />
      <button
        class="photon-button photon-button-micro"
        type="button"
      >
        Re-symbolicate profile
      </button>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Application
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Name and version:
      </span>
      Firefox
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Platform
  </h2>
  <div
    class="metaInfoSection"
  />
  <details
    class="metaInfoRow"
    open=""
  >
    <summary
      class="moreInfoSummary"
    >
      <button
        class="moreInfoButton photon-button photon-button-default photon-button-micro"
        type="button"
      >
        Show More
      </button>
    </summary>
    <div
      class="moreInfoPart"
    >
      <div>
        <h2
          class="metaInfoSubTitle"
        >
          CPU
        </h2>
        <div
          class="metaInfoSection"
        >
          <div
            class="moreInfoRow"
          >
            <span
              class="metaInfoWideLabel"
            >
              CPU 1
            </span>
            <div
              class="moreInfoValue"
            >
              Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
            </div>
          </div>
        </div>
      </div>
      <div>
        <h2
          class="metaInfoSubTitle"
        >
          Memory
        </h2>
        <div
          class="metaInfoSection"
        />
      </div>
      <div>
        <h2
          class="metaInfoSubTitle"
        >
          Hard Drives
        </h2>
        <div
          class="metaInfoSection"
        >
          <div
            class="moreInfoRow"
          >
            <span
              class="metaInfoWideLabel"
            >
              SSD
            </span>
            <div
              class="moreInfoValue"
            >
              Samsung SSD 850 EVO 500GB
            </div>
          </div>
          <div
            class="moreInfoRow"
          >
            <span
              class="metaInfoWideLabel"
            >
              HDD
            </span>
            <div
              class="moreInfoValue"
            >
              Seagate ST1000LM035-1RK172
            </div>
          </div>
        </div>
      </div>
    </div>
  </details>
</div>
`;
